### PR TITLE
Timr1126/xmeta model update

### PR DESCRIPTION
**Pull request by timr1126**
Restructure the xmeta models. This includes making fields private, adding accessor methods, restructuring the type references, and changing various field types in each model.

Each model holds its own `std::string` instead of a `std::string_view` so that the symbol table lifetime isn't dependent on the source data, which in the case of xmeta is the ANTLR AST.

### CODE REVIEW

File: src/tool/xmeta/cpp/models/model_types.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        void set_type_semantics(type_semantics const& sem)
+        {
+            m_semantic = sem;
+        }
+
+    private:
+        std::variant<std::string, type_semantics> m_semantic;
+    };
+}
```
Comment (kythant) [inline src/tool/xmeta/cpp/models/model_types.h]: nit: add new line here

File: src/tool/xmeta/cpp/models/enum_model.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    void enum_model::add_member(std::string_view const& id, std::optional<enum_member_semantics> val)
+    {
+        m_members.emplace_back(enum_member{ id, val });
+    }
+
+    void enum_model::add_member(std::string_view const& id, std::string_view const& expression_id)
+    {
+        m_members.emplace_back(enum_member{ id, expression_id });
+    }
+}
```
Comment (kythant) [inline src/tool/xmeta/cpp/models/enum_model.cpp]: nit: add new line here

File: src/tool/xmeta/cpp/models/method_model.h [hunk 3]
```diff
@@ -1,32 +1,68 @@
...
 {
-    struct method_modifiers
+    struct method_semantics
     {
-        bool is_protected;
-        bool is_override;
-        bool is_static;
-        bool is_sealed;
+        bool is_protected = false;
+        bool is_static = false;
```
Comment (kythant) [inline src/tool/xmeta/cpp/models/method_model.h]: I think it's abit more clear if there is is_override here.

File: src/tool/xmeta/cpp/models/using_directive_models.h [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        void set_namespace_ref(std::shared_ptr<namespace_model> const& nm)
+        {
+            m_namespace_ref = nm;
+        }
+
+    private:
+        std::variant<std::string, std::shared_ptr<namespace_model>> m_namespace_ref;
+    };
+}
```
Comment (kythant) [inline src/tool/xmeta/cpp/models/using_directive_models.h]: nit: add new line here

File: src/tool/xmeta/cpp/models/class_model.h [hunk 5]
```diff
@@ -13,27 +15,85 @@
...
+            m_static_interfaces.emplace_back(static_interface_ref);
+        }
+
+    private:
+        std::variant<std::string, std::shared_ptr<class_model>> m_class_base_ref;
+        std::vector<std::variant<std::string, std::shared_ptr<interface_model>>> m_interface_base_refs;
+        class_semantics m_semantic;
+        // TODO: Add type parameters (generic types)
+
+        std::vector<std::variant<
```
[the remark this answers is not in the document]
Comment (timr1126) [reply src/tool/xmeta/cpp/models/class_model.h]: Ah ok. I was under the impression the order of class members mattered for metadata generation.

[the remark this answers is not in the document]
Comment (DefaultRyan) [reply src/tool/xmeta/cpp/models/class_model.h]: Good point, you're right. Never mind then. :)

File: src/tool/xmeta/cpp/models/class_model.h [hunk 6]
```diff
@@ -13,27 +15,85 @@
...
+        }
+
+        auto const& get_semantic()
+        {
+            return m_semantic;
+        }
+
+        void add_interface_base_ref(std::string_view const& interface_base_ref)
+        {
+            m_interface_base_refs.emplace_back(std::string(interface_base_ref));
```
[the remark this answers is not in the document]
Comment (timr1126) [reply src/tool/xmeta/cpp/models/class_model.h]: `emplace_back` expects either a `std::string` or a `std::shared_ptr<interface_model>`. It won't accept a `std::string_view`. I can make the parameter a `std::string const&` instead.

[the remark this answers is not in the document]
Comment (DefaultRyan) [reply src/tool/xmeta/cpp/models/class_model.h]: No, I think the parameter type being `string_view` is correct. I just forgot about that little wrinkle in the string conversion rules. The code as-is seems correct.

**Pull request merged**